src/resolve-system-dependencies/resolve-system-dependencies.cc

```diff
@@ -0,0 +1,204 @@
+        swap_fat_arch(&arch, 1, end);
+        if(arch.cputype == CPU_TYPE_X86_64) {
+            fseek(obj_file, arch.offset, SEEK_SET);
+            return true;
+        }
+    }
+    return false;
+}
+
+std::set<std::string> runResolver(const Path & filename) {
+    FILE *obj_file = fopen(filename.c_str(), "rb");
+    uint32_t magic;
+    fread(&magic, sizeof(uint32_t), 1, obj_file);
+    fseek(obj_file, 0, SEEK_SET);
+    enum NXByteOrder endianness;
+    if(magic == 0xBEBAFECA) {
+        endianness = NX_BigEndian;
+        if(!seekMach64Blob(obj_file, endianness)) {
+            std::cerr << "Could not find any mach64 blobs in file " << filename << ", continuing..." << std::endl;
+            return std::set<string>();
+        }
+    }
+    struct mach_header_64 header;
+    fread(&header, sizeof(struct mach_header_64), 1, obj_file);
+    if(!(header.magic == MH_MAGIC_64 || header.magic == MH_CIGAM_64)) {
+        std::cerr << "Not a mach-o object file: " << filename << std::endl;
+        return std::set<string>();
+    }
+    std::set<string> libs;
+    for(uint32_t i = 0; i < header.ncmds; i++) {
+        struct load_command cmd;
+        fread(&cmd.cmd, sizeof(cmd.cmd), 1, obj_file);
+        fread(&cmd.cmdsize, sizeof(cmd.cmdsize), 1, obj_file);
+        switch(cmd.cmd) {
+            case LC_LOAD_DYLIB:
+            case LC_LOAD_UPWARD_DYLIB:
+            case LC_REEXPORT_DYLIB:
+                libs.insert(findDylibName(obj_file, cmd));
+                break;
+        }
+        fseek(obj_file, cmd.cmdsize - (sizeof(uint32_t) * 2), SEEK_CUR);
+    }
+    fclose(obj_file);
+    libs.erase(filename);
+    return libs;
+}
+
+bool isSymlink(const Path & path) {
+    struct stat st;
+    if(lstat(path.c_str(), &st))
+        throw SysError(format("getting attributes of path ‘%1%’") % path);
+
+    return S_ISLNK(st.st_mode);
+}
+
+Path resolveSymlink(const Path & path) {
+    char buf[PATH_MAX];
+    ssize_t len = readlink(path.c_str(), buf, sizeof(buf) - 1);
+    if(len != -1) {
+        buf[len] = 0;
+        return Path(buf);
+    } else {
+        throw SysError(format("readlink('%1%')") % path);
+    }
+}
+
+std::set<string> resolveTree(const Path & path, PathSet & deps) {
+    std::set<string> results;
+    if(deps.find(path) != deps.end()) {
+        return std::set<string>();
+    }
+    deps.insert(path);
+    for (auto & lib : runResolver(path)) {
+        results.insert(lib);
+        for (auto & p : resolveTree(lib, deps)) {
+            results.insert(p);
+        }
+    }
+    return results;
+}
+
+std::set<string> getPath(const Path & path) {
+    Path cacheFile = resolveCacheFile(path);
+    if(pathExists(cacheFile)) {
+        return readCacheFile(cacheFile);
+    }
+
+    std::set<string> deps;
+    std::set<string> paths;
+    paths.insert(path);
+
+    Path next_path = Path(path);
+    while(isSymlink(next_path)) {
+        next_path = resolveSymlink(next_path);
+        paths.insert(next_path);
+    }
```

Comment:
Every intermediate link needs to be known here; we'll get permission denied from sandbox-exec if one is missing.